types/chain_id_test.go

```diff
@@ -0,0 +1,113 @@
+package types
+
+import (
+	_ "embed"
+	"strings"
+	"testing"
+
+	genutiltypes "github.com/cosmos/cosmos-sdk/x/genutil/types"
+	"github.com/stretchr/testify/require"
+)
+
+//go:embed testdata/parse_chain_id.json
+var BenchmarkGenesis string
+
+func TestParseChainIDFromGenesis(t *testing.T) {
+	testCases := []struct {
+		name       string
+		json       string
+		expChainID string
+	}{
+		{
+			"success",
+			`{
+				"state": {
+				  "accounts": {
+					"a": {}
+				  }
+				},
+				"chain_id": "test-chain-id"
+			}`,
+			"test-chain-id",
+		},
+		{
+			"nested",
+			`{
+				"state": {
+				  "accounts": {
+					"a": {}
+				  },
+				  "chain_id": "test-chain-id"
+				}
+			}`,
+			"",
+		},
+		{
+			"not exist",
+			`{
+				"state": {
+				  "accounts": {
+					"a": {}
+				  }
+				},
+				"chain-id": "test-chain-id"
+			}`,
+			"",
+		},
+		{
+			"invalid type",
+			`{
+				"chain-id":1,
+			}`,
+			"",
+		},
+		{
+			"invalid json",
+			`[ " ': }`,
+			"",
+		},
+		{
+			"empty chain_id",
+			`{"chain_id": ""}`,
+			"",
+		},
+		{
+			"chain_id too long",
+			`{"chain_id": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"}`,
+			"",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			chain_id, err := ParseChainIDFromGenesis(strings.NewReader(tc.json))
+			if tc.expChainID == "" {
+				require.Error(t, err)
```

Comment:
If you wanna rigorously tighten the checks, you can do a substring match for the expected error e.g. require.Contains(t, err.Error(), "not a string")

Reply:
done, fixed.